content/blog/2019-10-18-double-checked-locking/index.md

```diff
@@ -0,0 +1,353 @@
+Some compilers, memory systems, or processors may reorder the instructions,
+     so that the initialization of the object is done after the assignment of the pointer to the object.
+     Another thread may see that the pointer is assigned and access the fields of the object before the fields of the object is initialized.
+
+### Another fix is also broken
+
+Some people came up with another fix for this issue with a memory barrier.
+A memory barrier is a type of instruction that can make the compiler and processor
+enforce the ordering, so that the instructions on one side of the memory barrier will
+not be reordered to the other side of the barrier.
+
+```Java
+class Foo {
+    private Helper helper = null;
+
+    public Helper getHelper() {
+        if (helper == null) {
+            Helper h;
+            synchronized(this) {
+                h = helper;
+                if (h == null)
+                    synchronized(this) {
+                        h = new Helper();
+                    }                       
+                helper = h;
+            }
+        }
+        return helper;
+    }
+}
+
+```
+
+The intuitive idea is that the lock releasing would act as a memory barrier,
+so that `helper=h` will not be executed until the initialization 
+in the synchronized section is done.
+
+Unfortunately, the lock releasing is a one-way memory barrier on many processors.
+It only enforces that the instructions in the synchronized section 
+must be executed before lock releasing.
+The instruction `helper=h` behind the memory barrier could still be moved into 
+synchronized section and executed before the object initialization is done.
+
+## Working Solutions
+
+
+### Volatile in JDK5
+
+Java introduced "volatile" since JDK5. 
+A "volatile" makes the variable be stored and accessed in main memory.
+Every read of a volatile is from main memory.
+Every write of a volatile is to main memory.
+
+The "volatile" also provides a "happen-before" guarantee:
+
+* The reads/writes of other variables before a write to a volatile variable 
+cannot be reordered after the write to the volatile variable.
+* The reads/writes of other variables after a read from a volatile variable
+cannot be reordered before the read from the volatile variable.
+
+With this new feature, the double-checked locking issue is resolved by simply changing
+the `helper` to volatile, since it guarantees that the initialization of the
+object `new Helper()` happens before the assignment to the `helper`.
+
+```Java
+class Foo {
+    private volatile Helper helper = null;
+    public Helper getHelper() {
+        if (helper == null) {
+            synchronized(this) {    
+                if (helper == null)
+                    helper = new Helper();
+            }
+        }
+        return helper;
+    }
+}
+```
+
+However, since all read and write operations of a volatile variable need to
+access main memory, which can be very slow. 
```

Comment:
Again, it's not just that they "access main memory"—they also use special hardware support for atomic/synchronized accesses (i.e., the compiler inserts barriers anyway), so nothing is really gained.